### PR TITLE
feature(Consents): Users should see what they have consented to.

### PR DESCRIPTION
**Pull request by lauryndbrown**
Fixes https://github.com/carpentries/amy/issues/1830.
Dependent upon: https://github.com/carpentries/amy/pull/1832

### CODE REVIEW

**Comment by lauryndbrown** [comment]
Currently looks like this:
Screen Shot 2021-03-10 at 7 21 49 AM

**Comment by lauryndbrown** [comment]
Not the most restful, and a little hacky. Need to revise.

**Comment by lauryndbrown** [comment]
This approach runs into problems when we want to delete consents. But could be done. We would simply have to add a widthdraw consent option.

A table might be better here, showing:
- consent name
- date consented
- required for the profile
- edit button (blue)
- withdraw button (red)

I think I like the table a bit better, since it's more consistent with the rest of the application, but I will still need to build the options for the form no matter what happens, just given the way the code is written. There's no need to recreate the form's display logic.

**Comment by lauryndbrown** [comment]
@pbanaszkiewicz could you take a look for me?

**Comment by pbanaszkiewicz** [comment]
@lauryndbrown I will into this one, probably later tonight.

**Review by pbanaszkiewicz — COMMENTED** [review]
This is looking okay; creating the form on-the-fly is interesting approach and I think you did a great job there.

Personally I think I'd try first working with formsets and see if that takes me somewhere.

File: amy/consents/forms.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+from typing import Iterable
+from consents.models import Consent
+from django import forms
+from workshops.forms import BootstrapHelper
+from consents.models import Term, TermOption
+from workshops.forms import WidgetOverrideMixin
+from workshops.models import Person
+from django.utils import timezone
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply amy/consents/forms.py:8]: Oh gotcha. I think I might be used  to isort: https://pypi.org/project/isort/ a tool that sorts and groups them for me. I'll be sure to double check before.Maybe run it locally if we don't want to add something like that.

File: amy/consents/forms.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                consent.archived_at = timezone.now()
+                consent.save()
+            Consent.objects.create(
+                person=person, term_option_id=option_id, term_id=term.id
+            )
+
+    def _yes_only(self, term) -> bool:
+        return len(term.options) == 1
+
+    def _yes_and_no(self, term) -> bool:
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply amy/consents/forms.py:89]: I was going to change the display of the yes only and yes and no options. I think a check box would probably work better. But I'll leave them the same for now. And remove that.

File: amy/consents/forms.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        Construct a Form of all nonarchived Terms with the
+        consent answers added as initial.
+        """
+        self.terms = Term.objects.active().prefetch_active_options()
+        term_id_by_option_id = {
+            consent.term_id: consent.term_option_id
+            for consent in Consent.objects.filter(
+                archived_at=None, term__in=self.terms, person=person
+            )
+        }
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply amy/consents/forms.py:45]: So here there are only two queries occurring:
- `Term.objects.active().prefetch_active_options()`
- `Consent.objects.filter(archived_at=None, term__in=self.terms, person=person)`

The use of `consent.term_id: consent.term_option_id` by using `_id` I am referencing the forignkey id in the `Consent` table. And no additional query is needed.

File: amy/consents/views.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from consents.models import Consent
+from workshops.base_views import RedirectSupportMixin
+from consents.forms import ConsentForm
+from workshops.base_views import AMYCreateView
+
+
+class ConsentUpdate(RedirectSupportMixin, AMYCreateView):
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply amy/consents/views.py]: @pbanaszkiewicz I didn't see any usage of `LoginRequiredMixin` in the code. Do you typically use something else?

**Comment by lauryndbrown** [comment]
@pbanaszkiewicz I think I handled most of the issues. Could you take another look?

File: amy/consents/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pbanaszkiewicz) [reply amy/consents/views.py]: We're using `@login_required` for function-based views. The CBV equivalent is `LoginRequiredMixin`. We have a bunch of decorators and mixins for this purpose in `workshops.util`.

File: amy/consents/forms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pbanaszkiewicz) [reply amy/consents/forms.py:5]: Yes, I was thinking about adding flake8 and isort to the precommit-hook, but didn't get to it yet.

File: amy/consents/forms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pbanaszkiewicz) [reply amy/consents/forms.py:49]: Perfect, then no additional prefetching is needed.

**Review by pbanaszkiewicz — APPROVED** [review]
Looks good @lauryndbrown !

**Pull request merged**